File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp
class Solution {
public:

    int dx[4] = {1,-1,0,0};
    int dy[4] = {0,0,1,-1};

    bool isValid(int x,int y,vector<vector<int>> &vis){
        if( x< 0 || y<0 || x>=vis.size() || y>= vis[0].size())
            return false;

        return true;    
    }
    void bfs(int i,int j, vector<vector<int>> &vis,vector<vector<char>>& grid){

        queue<pair<int,int>> q;

        q.push({i,j});
        vis[i][j]=1;
        int x,y;
        while(q.size()>0){

            auto it = q.front();
            q.pop();
            for(int k=0;k< 4 ;k++){
                x = dx[k]+it.first;
                y = dy[k]+it.second;

                if(isValid(x,y,vis)){
                    if(vis[x][y]==0 && grid[x][y]=='1'){
                        q.push({x,y});
                        vis[x][y]=1;
                    }
                }
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        vector<vector<int>> vis(m,vector<int> (n,0));
        int res=0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(vis[i][j]==0 && grid[i][j]=='1'){
                    bfs(i,j,vis,grid);
                    res++;
                }
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-0.java
class Solution {

    int[] dx = {1,0,0,-1};
    int[] dy = {0,1,-1,0};

    boolean isValid(int i,int j,int n,int m){
        return !(i<0 || j<0 || i>=n || j>=m);
    }
    private int dfs(int [][]matrix, int [][]vis,int i,int j){
        if(vis[i][j]!=0)
            return vis[i][j];

        int x,y;
        int res=0;
        int n = matrix.length,  m = matrix[0].length;

        for(int k=0;k<4;k++){

            x = i + dx[k];
            y = j + dy[k];

            if(isValid(x,y,n,m)){
                if(matrix[x][y] > matrix[i][j]){
                    if(vis[x][y] ==0)
                        vis[x][y] = dfs(matrix,vis,x,y);
                    
                    res = Math.max(res,vis[x][y]);
                }
            }
            
        }
        return vis[i][j] = 1 + res;
    }
    public int longestIncreasingPath(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        int[][] dp= new int [n][m];

        for(int i=0;i<n;i++)
            for(int j=0;j<m;j++)
                dp[i][j]=0;

        int res=0;
        for(int i=0;i<n;i++)
            for(int j=0;j<m;j++)
                if(dp[i][j]==0)
                    res = Math.max(res,dfs(matrix,dp,i,j));     


        return res;

    }
}
